ksqldb-execution/src/main/java/io/confluent/ksql/execution/codegen/CodeGenRunner.java

```diff
@@ -175,20 +176,33 @@ private Visitor() {
 
     @Override
     public Void visitLikePredicate(final LikePredicate node, final TypeContext context) {
-      process(node.getValue(), null);
-      process(node.getPattern(), null);
+      process(node.getValue(), context);
+      process(node.getPattern(), context);
       return null;
     }
 
     @Override
     public Void visitFunctionCall(final FunctionCall node, final TypeContext context) {
       final List<SqlArgument> argumentTypes = new ArrayList<>();
       final FunctionName functionName = node.getName();
+      final boolean hasLambda = node.hasLambdaFunctionCallArguments();
       for (final Expression argExpr : node.getArguments()) {
-        process(argExpr, null);
-        argumentTypes.add(SqlArgument.of(
-            expressionTypeManager.getExpressionSqlType(argExpr)
-        ));
+        final TypeContext childContext = context.getCopy();
+        final SqlType resolvedArgType =
+            expressionTypeManager.getExpressionSqlType(argExpr, childContext);
+        process(argExpr, context.getCopy());
+
+        if (argExpr instanceof LambdaFunctionCall) {
+          argumentTypes.add(
+              SqlArgument.of(
+                  SqlLambda.of(context.getLambdaInputTypes(), childContext.getSqlType())));
```

Comment:
nit: I think this is cleaner to just use `resolvedArgType` as the second parameter, which will be the same as `childContext.getSqlType()`. Ditto on other two classes.

Reply:
Also maybe we could consolidate the shared logic in `FunctionCall`:

```
List<SqlArgument> argumentTypes resolveArgumentTypes(TypeContext context)
```

shared among `CodeGenRunner`, `SqlToJavaVisitor` and `ExpressionTypeManager`?

And I'm even wondering, why we need to try to resolve the argument types multiple times, rather than just cache the resolved types inside the `FunctionCall` node itself after done for the first time, and then re-use it in later iterations of traversals?

Reply:
Agreed @guozhangwang I think we can definitely optimize these 3 classes to avoid duplicate traversals. Thanks for filing the issue. I think for now since this was an existing issue even before lambdas, it's out of scope for the lambdas implementation and we can follow up on issue you filed in the future.